pkg/provider/aws/resources/vpc.go

```diff
@@ -88,6 +88,230 @@ type (
+//
+// This method will also create dependent resources which are necessary for functionality. Those resources are:
+//   - Subnet
+//   - Elastic IP
+func (nat *NatGateway) Create(dag *core.ResourceGraph, params NatCreateParams) error {
+
+	nat.Name = natGatewaySanitizer.Apply(fmt.Sprintf("%s-%s", params.AppName, params.AZ))
+	nat.ConstructsRef = params.Refs
+
+	existingNat := dag.GetResourceByVertexId(nat.Id().String())
+	if existingNat != nil {
+		graphNat := existingNat.(*NatGateway)
+		graphNat.ConstructsRef = append(graphNat.ConstructsRef, params.Refs...)
+	} else {
+		subResourceParams := map[string]any{
+			"Subnet": SubnetCreateParams{
+				AppName: params.AppName,
+				Refs:    params.Refs,
+				AZ:      params.AZ,
+				Type:    PublicSubnet,
+			},
+			"ElasticIp": EipCreateParams{
+				AppName: params.AppName,
+				Refs:    params.Refs,
+				IpName:  params.AZ,
+			},
+		}
+		err := dag.CreateDependencies(nat, subResourceParams)
+		return err
+	}
+	return nil
+}
+
+type SubnetCreateParams struct {
+	AppName string
+	Refs    []core.AnnotationKey
+	AZ      string
+	Type    string
+}
+
+// Create takes in an all necessary parameters to generate the Subnet name and ensure that the Subnet is correlated to the constructs which required its creation.
+//
+// This method will also create dependent resources which are necessary for functionality. Those resources are:
+//   - VPC
+//   - Route Table
+//   - Nat Gateway (if subnet is private)
+//   - Internet Gateway (if subnet is private)
+func (subnet *Subnet) Create(dag *core.ResourceGraph, params SubnetCreateParams) error {
+	subnet.Name = subnetSanitizer.Apply(fmt.Sprintf("%s-%s%s", params.AppName, params.Type, params.AZ))
+	subnet.ConstructsRef = params.Refs
+	subnet.AvailabilityZone = core.IaCValue{Resource: NewAvailabilityZones(), Property: params.AZ}
+	subnet.Type = params.Type
+
+	routeTableParams := RouteTableCreateParams{
+		AppName: params.AppName,
+		Refs:    params.Refs,
+	}
+	if subnet.Type == PrivateSubnet {
+		routeTableParams.Name = fmt.Sprintf("%s%s", params.Type, params.AZ)
+	} else {
+		routeTableParams.Name = fmt.Sprintf(params.Type)
+	}
+	rt := &RouteTable{}
+	err := rt.Create(dag, routeTableParams)
+	if err != nil {
+		return err
+	}
+	if subnet.Type == PrivateSubnet {
+		nat := &NatGateway{}
+		natParams := NatCreateParams{
+			AppName: params.AppName,
+			Refs:    params.Refs,
+			AZ:      params.AZ,
+		}
+		err := nat.Create(dag, natParams)
+		if err != nil {
+			return err
+		}
+		dag.AddDependency(rt, nat)
+	} else if subnet.Type == PublicSubnet {
+		igw := &InternetGateway{}
+		igwParams := IgwCreateParams{
+			AppName: params.AppName,
+			Refs:    params.Refs,
+		}
+		err := igw.Create(dag, igwParams)
+		if err != nil {
+			return err
+		}
+		dag.AddDependency(rt, igw)
+	}
+
+	err = dag.CreateDependencies(subnet, map[string]any{
```

Comment:
Shouldn't the VPC be created before all the gateways, routetables, etc? If frontloaded, could also save some creation time for cache hits.

Reply:
the vpc wouldnt get created explicitly, but we should be adding caching to the createDependencies calls to speed things up, just figured we could add that in later